Src/Workspaces/Core/Portable/Utilities/TaskExtensions.cs

```diff
@@ -337,5 +338,137 @@ public static T WaitAndGetResult<T>(this Task<T> task, CancellationToken cancell
                 cancellationToken, taskContinuationOptions, scheduler).Unwrap();
         }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the task.</typeparam>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        public static Task<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
+        {
+            if (task == null)
+            {
+                throw new ArgumentNullException(nameof(task));
+            }
+
+            if (!cancellationToken.CanBeCanceled || task.IsCompleted)
+            {
+                return task;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return SingletonTask<T>.CanceledTask;
+            }
+
+            return WithCancellationSlow(task, cancellationToken);
+        }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        public static Task WithCancellation(this Task task, CancellationToken cancellationToken)
+        {
+            if (task == null)
+            {
+                throw new ArgumentNullException(nameof(task));
+            }
+
+            if (!cancellationToken.CanBeCanceled || task.IsCompleted)
+            {
+                return task;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return SingletonTask<int>.CanceledTask;
+            }
+
+            return WithCancellationSlow(task, cancellationToken);
+        }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the task.</typeparam>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        private static async Task<T> WithCancellationSlow<T>(Task<T> task, CancellationToken cancellationToken)
```

Comment:
@CyrusNajmabadi The non generic version of this is alsy an *async* method and therefore returns an instance of `Task` not a `Task<T>`. So we can't cast it back in order to get its result. We have to implement it ourselves here.

@jasonmalinowski I think your proposal would work fine if `task` completed successfully or if `cancellationToken` is canceled. But if `task` transitioned to either a canceled or faulted state, the result would not be what the caller expects. If `task` faults, `t.Result` will throw AggregateException, then the task returned from `ContinueWith` will wrap that in _another_ AggregateException. So you end up wrapping the exception in extra layers that a catch block later may not be expecting. And if `task` transitions to a canceled state, instead of WithCancellation returning a canceled Task, it would return a `Task` faulted with an AggregateException and an inner `OperationCanceledException`. These methods are implemented to preserve behavior of the original task exactly when it completes. So yes, your way is simpler, but it doesn't meet the behavior requirement.